### PR TITLE
Fix deepcopy of user data

### PR DESCRIPTION
**Pull request by mgunyho**
#### Description

I was making use of the `user_data` attribute of a `Parameter`, and noticed that it is not deep-copied when using the `copy()` method of `Parameters`. Here is a simple example:
```python
import lmfit

def func(x, a):
    return a * x

model = lmfit.Model(func)

params = model.make_params()
params.get("a").user_data = {"foo": 123 }

params2 = params.copy()
params2.get("a").user_data["foo"] = 456

print(params.get("a").user_data) # prints {'foo': 456}
```

This PR fixes this behavior.

###### Type of Changes

- [x] Bug fix
- [ ] New feature
- [ ] Refactoring / maintenance
- [ ] Documentation / examples

###### Tested on

Python: 3.10.9 (main, Dec 08 2022, 14:49:06) [GCC]

lmfit: 1.1.0.post1+g1a3b1009, scipy: 1.10.0, numpy: 1.24.1, asteval: 0.9.28, uncertainties: 3.1.7

###### Verification 
Have you

- [ ] included docstrings that follow PEP 257? - N/A

- [ ] referenced existing Issue and/or provided relevant link to mailing list? - N/A

- [x] verified that existing tests pass locally?

- [ ] verified that the documentation builds locally? - N/A, no changes to docs

- [x] squashed/minimized your commits and written descriptive commit messages? - let me know if you want me to squash them into one commit

- [x] added or updated existing tests to cover the changes?
- [ ] updated the documentation and/or added an entry to the release notes (doc/whatsnew.rst)?
- [ ] added an example? - N/A

### CODE REVIEW

File: tests/test_parameters.py [hunk 1]
```diff
@@ -20,9 +20,9 @@ def parameters():
                              user_data=1))
     pars.add(lmfit.Parameter(name='b', value=0.0, vary=True, min=-250.0,
                              max=250.0, expr="2.0*a", brute_step=25.0,
-                             user_data=2.5))
+                             user_data={'test': 123}))
     exp_attr_values_A = ('a', 10.0, True, -100.0, 100.0, None, 5.0, 1)
-    exp_attr_values_B = ('b', 20.0, False, -250.0, 250.0, "2.0*a", 25.0, 2.5)
+    exp_attr_values_B = ('b', 20.0, False, -250.0, 250.0, "2.0*a", 25.0, {'test': 123})
```
Comment (reneeotten) [inline tests/test_parameters.py:25]: what's the purpose of this change - why would one want to make things fail when `user_data` is a dictionary? Or am I missing something here?

File: lmfit/parameter.py [hunk 2]
```diff
@@ -109,7 +109,7 @@ def __deepcopy__(self, memo):
                 param.correl = par.correl
                 param.init_value = par.init_value
                 param.expr = par.expr
-                param.user_data = par.user_data
+                param.user_data = deepcopy(par.user_data)
```
Comment (reneeotten) [inline lmfit/parameter.py:112]: I don't have a real issue with doing this, but why is it explicitly needed here and not for any of the other attributes - that doesn't make complete sense to me, but admittedly I haven't looked carefully yet...

File: tests/test_parameters.py (hunk 1 above)
Comment (mgunyho) [reply tests/test_parameters.py:25]: With this change, the test covers a case that was not covered before, where the user data is a mutable object such as a dict or list. See the example code in the PR: modifying the user data of the copied `params2` modifies the user data of the original `params`.

File: lmfit/parameter.py (hunk 2 above)
Comment (mgunyho) [reply lmfit/parameter.py:112]: I am quite new to lmfit, so I'm not sure which of the other attributes are mutable types, maybe `correl` is for example. I agree that it looks a bit asymmetric.

**Comment by newville** [comment]
@mgunyho @reneeotten  Thanks - I do agree that `user_data` is the most likely to be mutable.   I think we should do a `deepcopy` on `user_data` and `correl`  and probably everything in the `user_defined_symbols`, as with
   ```
unique_symbols = {key: deepcopy(self._asteval.symtable[key]) for key in self._asteval.user_defined_symbols()}
```

**Comment by mgunyho** [comment]
That seems reasonable. I am not familiar with the internals of `Parameter`, and I'm not sure if/how the tests should be updated to check that the behavior wrt `_asteval` or the other attributes is correct. You should be able to add commits to this PR, so feel free to do so if you don't want to make a separate one.

File: tests/test_parameters.py (hunk 1 above)
Comment (reneeotten) [reply tests/test_parameters.py:25]: I see what you mean. In the current test "user_data" was an int/float value in both cases, replacing one with a more complicated type makes sense!

**Comment by newville** [comment]
@mgunyho @reneeotten will merge and update #844
